### ISSUE TITLE
[Arrow functions] { return; } is typed as Dynamic rather than Void

### ISSUE THREAD

**Issue by haxiomic** [HaxeFoundation/haxe#7890]
**TL;DR**: ` { return; }` gets typed as Dynamic, but something like `{ trace('hi'); }` gets typed as Void. This creates unexpected errors when using arrow functions that you expect to have a Void return type.

A specific example, I'm having an issue using arrow functions with `js.Promise`:
http://try-haxe.mrcdk.com/#7d772

```haxe
var p = js.Promise.resolve('hi');

p.then(
         // this method has type String -> Dynamic
	(string) -> {
		// do nothing, this would be expected to make the function type `String -> Void`
		return;
	},
	(error) -> {
		trace(error);
	}
);
```

This unexpectedly fails with
```
Test.hx:8: lines 8-11 : Dynamic should be Void
Test.hx:8: lines 8-11 : For function argument 'onFulfilled'
```

It fails because the the second function is supposed to have the same return type as the first but the first has return type Dynamic whereas the second has return type Void.

I'm not sure `{ return; }` should type as Dynamic, especially when  `function() { return; }` has type `() -> Void`

**Comment by nadako** [comment]
This is funny. `(string) -> { return; }` is the same as `function(string) return { return; }`, and the `return` expression itself is a monomorph so stuff like `var x = if (...) 1 else return;` works. I think we can be smarter here somehow, but I'm not sure how exactly right now.

**Comment by haxiomic** [comment]
Nice one @Simn! Thanks for such a quick turn around - I was scratching my head looking the typer code thinking this could be a tricky issue to fix

**Comment by Simn** [comment]
This is going to require some more investigation, see commit comments.
